tempesta_fw/apm.c

```diff
@@ -299,47 +326,51 @@ tfw_stats_adjust(TfwPcntRanges *rng, int r)
  * We only care about correct array indexing.
  */
 static void
-tfw_stats_update(TfwPcntRanges *rng, unsigned int r_time)
+tfw_stats_update(TfwPcntRanges *rng, unsigned int r_time, spinlock_t *slock)
 {
 	TfwPcntCtl pc3, pc2 = { .atomic = rng->ctl[2].atomic };
 
+	/* Adjust min/max values. */
+	if (!tfw_stats_adj_min(rng, r_time))
+		tfw_stats_adj_max(rng, r_time);
+	/* Add to @tot_val for AVG calculation. */
+	atomic64_add(r_time, &rng->tot_val);
```

Comment:
That's dangerous. While we have 64 bits, each small delta-t there are many requests on the fly processed at many CPU cores at the same time, so each small deltat-t we're adding relatively large values (request processing latency) to the counter on all the CPU cores.

Rather we should use formula:

       avg = ((avg * tot_n) + r_time) / ++tot_n

, where `tot_n` is total number of requests. Since we're working in multi core environment we should calculate the statistic per cpu to avoid races in counters. I think `min` and `max` also are good to be calculated per cpu to about the loops over atomic caunters. Thus, procfs interface can show average in following way:

        avg = (cpu0->avg * cpu0->tot_n + ... + cpuN->avg * cpuN->tot_n) / (cpu0->tot_n + ... + cpuN->tot_n)

Calculation of total `min` and `max` is trivial.